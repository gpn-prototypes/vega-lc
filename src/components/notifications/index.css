.LC_Notifications {
  position: absolute;
  z-index: 2;
  left: 50%;
  bottom: 0;

  width: 400px;

  transform: translate(-50%, 0);

  &__SnackBar {
    padding-right: var(--space-xs);
    padding-left: var(--space-xs);

    overflow: auto;

    width: 400px;
  }
}
