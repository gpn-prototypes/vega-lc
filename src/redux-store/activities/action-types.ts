import { applyActionTypesNames } from '@/utils/apply-action-types-names';

export type ActivitiesActions =
  | 'SET_ACTIVITIES_LIST'
  | 'SET_SEARCH_STRING'
  | 'SET_IS_AUTO_FOCUS'
  | 'SET_ACTIVITIES_REF'
  | 'SET_DRAGGING_ELEMENTS'
  | 'SET_IS_DROPPING_ON_EXISTING_STEP';

const ActivitiesActionTypes: Record<ActivitiesActions, string> = {
  SET_ACTIVITIES_LIST: '',
  SET_SEARCH_STRING: '',
  SET_IS_AUTO_FOCUS: '',
  SET_ACTIVITIES_REF: '',
  SET_DRAGGING_ELEMENTS: '',
  SET_IS_DROPPING_ON_EXISTING_STEP: '',
};

applyActionTypesNames(ActivitiesActionTypes, 'Activities');

export { ActivitiesActionTypes };
