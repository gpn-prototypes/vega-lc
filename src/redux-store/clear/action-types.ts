import { applyActionTypesNames } from '@/utils/apply-action-types-names';

export type ClearActions = 'LC_CLEAR_STORES';

const ClearActionTypes: Record<ClearActions, string> = {
  LC_CLEAR_STORES: '',
};

applyActionTypesNames(ClearActionTypes, 'Clear');

export { ClearActionTypes };
