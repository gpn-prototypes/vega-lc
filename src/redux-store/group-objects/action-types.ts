import { applyActionTypesNames } from '../../utils/apply-action-types-names';

const GroupObjectsActionTypes = {
  SET_GROUP_OBJECTS_LIST: null,
};

applyActionTypesNames(GroupObjectsActionTypes, 'GroupObjects');

export { GroupObjectsActionTypes };
