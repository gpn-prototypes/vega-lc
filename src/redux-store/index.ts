import { applyMiddleware, combineReducers, createStore, Store } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension/developmentOnly';
import { persistStore } from 'redux-persist';
import thunkMiddleware from 'redux-thunk';

import activitiesReducer from './activities/reducer';
import groupObjectsReducer from './group-objects/reducer';
import logicConstructorReducer from './logic-constructor/reducer';
import notificationsReducer from './notifications/reducer';
import projectStructureReducer from './project-structure/reducer';
import versionReducer from './version/reducer';

import { StoreLC } from '@/types/redux-store';

const rootReducer = combineReducers({
  activities: activitiesReducer,
  groupObjects: groupObjectsReducer,
  projectStructure: projectStructureReducer,
  logicConstructor: logicConstructorReducer,
  version: versionReducer,
  notifications: notificationsReducer,
});

const getStore = (initialState?: StoreLC): Store =>
  createStore(
    rootReducer,
    initialState,
    composeWithDevTools(applyMiddleware(thunkMiddleware.withExtraArgument({}))),
  );

export const store = getStore();

export const persistor = persistStore(store);
