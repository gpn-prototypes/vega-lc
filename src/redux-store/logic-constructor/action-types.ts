import { applyActionTypesNames } from '../../utils/apply-action-types-names';

type LogicConstructorActions =
  | 'SET_SCENARIO_LIST'
  | 'SET_CANVAS_ELEMENTS'
  | 'ADD_CANVAS_ELEMENT'
  | 'CREATE_STEP'
  | 'UPDATE_STEP'
  | 'SYNC_CANVAS_STATE'
  | 'TOGGLE_STEP_EDITOR'
  | 'SET_CANVAS_VIEW_REF';

const LogicConstructorActionTypes: Record<LogicConstructorActions, string> = {
  SET_SCENARIO_LIST: '',
  SET_CANVAS_ELEMENTS: '',
  ADD_CANVAS_ELEMENT: '',
  CREATE_STEP: '',
  UPDATE_STEP: '',
  SYNC_CANVAS_STATE: '',
  TOGGLE_STEP_EDITOR: '',
  SET_CANVAS_VIEW_REF: '',
};

applyActionTypesNames(LogicConstructorActionTypes, 'LogicConstructor');

export { LogicConstructorActionTypes };
