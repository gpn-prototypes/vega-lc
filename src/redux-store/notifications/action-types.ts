import { applyActionTypesNames } from '../../utils/apply-action-types-names';

type NotificationsActions = 'SET_NOTIFICATION' | 'DELETE_NOTIFICATION';

const VersionActionTypes: Record<NotificationsActions, string> = {
  SET_NOTIFICATION: '',
  DELETE_NOTIFICATION: '',
};

applyActionTypesNames(VersionActionTypes, 'Notification');

export { VersionActionTypes };
