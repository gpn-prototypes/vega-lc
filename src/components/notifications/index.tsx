import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Item } from '@consta/uikit/SnackBar';
import { SnackBar } from '@gpn-prototypes/vega-ui';

import { cnNotifications } from './cn-notifications';

import './index.css';

import { deleteNotification } from '@/redux-store/notifications/actions';
import { Notification, StoreLC } from '@/types/redux-store';

export const Notifications = (): React.ReactElement => {
  const dispatch = useDispatch();

  const notifications = useSelector<StoreLC, Notification[]>((state) => state.notifications.list);

  const items = notifications.map(
    (notification, index) =>
      ({
        key: index,
        message: notification.message,
        status: notification.status,
        icon: notification.icon,
        onClose: () => {
          dispatch(deleteNotification(index));
        },
      } as Item),
  );

  return (
    <div className={cnNotifications()}>
      <SnackBar className={cnNotifications('SnackBar')} items={items} />
    </div>
  );
};
