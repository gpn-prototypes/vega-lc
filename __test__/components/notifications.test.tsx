import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import { fireEvent, render, waitFor } from '@testing-library/react';
import { Store } from 'redux';
import configureMockStore from 'redux-mock-store';
import ResizeObserver from 'resize-observer-polyfill';

import '../../src/types/global';

import { Notifications } from '../../src/components/notifications';
import { VersionActionTypes } from '../../src/redux-store/notifications/action-types';
import initialState from '../../src/redux-store/notifications/initial-state';
import { Notification } from '../../src/types/redux-store';

const mockStore = configureMockStore();

const renderComponent = (store: Store) => {
  return render(
    <Provider store={store}>
      <Notifications />
    </Provider>,
  );
};

describe('Notifications', () => {
  test('рендерится без ошибок', () => {
    window.ResizeObserver = ResizeObserver;

    const storeData = {
      notifications: { ...initialState },
    };

    const store = mockStore(storeData);

    const dom = renderer
      .create(
        <Provider store={store}>
          <Notifications />
        </Provider>,
      )
      .toJSON();
    expect(dom).toMatchSnapshot();
  });

  test('отображаются нотификации', () => {
    const mockNotifications: Notification[] = [
      {
        message: 'mock notification 1',
        status: 'normal',
      },
      {
        message: 'mock notifications 2',
        status: 'normal',
      },
    ];

    const storeData = {
      notifications: { ...initialState, list: mockNotifications },
    };

    const store = mockStore(storeData);

    const component = renderComponent(store);

    const notifications = component.container.querySelectorAll('.SnackBar-Item');
    expect(notifications.length).toBe(2);

    const firstNotification = component.queryByText('mock notification 1');
    expect(firstNotification).toBeDefined();

    const secondNotification = component.queryByText('mock notification 2');
    expect(secondNotification).toBeDefined();
  });

  test('корректный статус у отображаемой нотификации', () => {
    const mockNotifications: Notification[] = [
      {
        message: 'normal notification',
        status: 'normal',
      },
      {
        message: 'alert notification',
        status: 'alert',
      },
      {
        message: 'success notification',
        status: 'success',
      },
    ];

    const storeData = {
      notifications: { ...initialState, list: mockNotifications },
    };

    const store = mockStore(storeData);

    const component = renderComponent(store);

    const notifications = component.container.querySelectorAll('.SnackBar-Item');
    expect(notifications.length).toBe(3);

    expect(notifications[0]).toHaveClass('SnackBar-Item_status_normal');
    expect(notifications[1]).toHaveClass('SnackBar-Item_status_alert');
    expect(notifications[2]).toHaveClass('SnackBar-Item_status_success');
  });

  test('вызывается функция удаления нотификации', async () => {
    const mockNotifications: Notification[] = [
      {
        message: 'normal notification',
        status: 'normal',
      },
    ];

    const storeData = {
      notifications: { ...initialState, list: mockNotifications },
    };

    const store = mockStore(storeData);

    const component = renderComponent(store);

    const notifications = component.container.querySelectorAll('.SnackBar-Item');
    expect(notifications.length).toBe(1);

    const closeButton = component.container.querySelector('.SnackBar-CloseButton') as HTMLElement;
    fireEvent.click(closeButton);

    const action = {
      type: VersionActionTypes.DELETE_NOTIFICATION,
      index: 0,
    };

    await waitFor(() => expect(store.getActions()).toContainEqual(action));
  });
});
