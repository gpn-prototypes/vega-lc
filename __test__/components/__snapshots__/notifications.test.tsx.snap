// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notifications рендерится без ошибок 1`] = `
<div
  className="LC_Notifications"
>
  <div
    className="SnackBar"
  />
</div>
`;
